Use Source Directory Structure

passle-chris opened the pull request:
I've made some modifications to allow an extra flag to be passed that when set to true will make the copy process create a directory structure in the target location. The directory structure will mimic the directory structure where the .ts file lived.

eg if the source project has the following:

Scripts\Components\Module1.ts
Scripts\Pages\Page1.ts

And the destination path is set to

C:\projects\MyProject\

Then it will copy the files to:
C:\projects\MyProject\Scripts\Components\Module1.js
C:\projects\MyProject\Scripts\Pages\Page1.js

The reason behind this change is that if a file has a reference to another typescript file that is in a different directory then when copied these references will no longer be valid. 

Similar if it references .d.ts files that are not copied then these references break. The easiest way to have these references work is to have the type files in your destination in the same relative location and this requires a certain amount of preservation of the directory structure.

The code changes are relatively simple and described in the commit in detail. Broadly though a new property will determine whether to use the destination directory as is or run some logic to find the project root relative path of the source and then append that path fragment to the destination. Default behaviour is to not use the new functionality (ie default behaviour will be as per the previous version).

Let me know if you have any questions with any of this.

passle-chris committed:
Merge pull request #1 from Zoltu/master

Merge pull request #9 from passle-chris/master

passle-chris committed:
Added in a new property "UseSourceDirectoryStructure". If this is set to true then the code will rather than copying directly to LibraryDirectoryFullPath will add in the folder structure where the typescript was found. So if the typescript is in a "Script/Module1" folder inside the typescript project then it will be copied to a Script/Module1 folder inside the target directory. This is useful for maintaining structure, for if you have multiple typescript files with the same name and if you want to preserve a folder structure for other reasons (such as to make sure references between typescript files will stay working as you expect).

To do this the LINQ to get the typescript was split up. We now enumerate through each project first so we can get the source Project's path so that we can then strip that off of the beginning of sourcedirectory path to get the root relative directory path. The main difference once inside the inner enumeration is that rather than using LibraryDirectoryFullPath property directly a method determines what the destination directory should be based on the value of the new property.

MicahZoltu commented:
I haven't used this package in a while but it sounds like this was just a bug and no setting is required.  It's there some reason why a user would want his file structure flattened?

MicahZoltu commented:
It looks like the build is falling.  Code contract it appears, usually this just means a missing null check.

MicahZoltu commented on code/FromReferencesTask.cs:
```diff
@@ -37,10 +37,20 @@ public bool CopyAll
 		{
 			try
 			{
+				System.Diagnostics.Debugger.Launch();
```
Did you mean to leave this in?

passle-chris replied:
Totally not!

MicahZoltu commented on code/FromReferencesTask.cs:
```diff
@@ -54,29 +64,39 @@ public override Boolean Execute()
-					.NotNull();
+				IEnumerable<ProjectRootElement> referencedProjects = GetReferencedProjects(project)
+									.NotNullToNull()
+									.Skip(1);
```
Style: chained method calls should be intended one more level from previous line.

MicahZoltu commented on code/FromReferencesTask.cs:
```diff
@@ -54,29 +64,39 @@ public override Boolean Execute()
+									.Skip(1);
 
-				foreach (var typeScriptFullPath in typeScriptFullPaths)
+				foreach (ProjectRootElement referencedProject in referencedProjects)
```
Style: recommend `var` here.

passle-chris commented:
With regards to whether this is a bug or a feature I can't think of a reason why you would want your file structure flattened but at the same time this is existing functionality and I didn't want peoples existing use of it to break due to changes to the package. I'm happy to remove the switch (or change the default around so it preserves file structure by default).

Also will go through and fix the issues you identified. Sorry I failed to notice the build failing. I'm still new to github so didn't think to look. Its all too clever for me. ;-)

MicahZoltu commented:
The style comments are not blockers for getting this PR merged, but would be nice to have changed.  Once the build is passing and the other comments addressed we can get this merged.

MicahZoltu commented:
If you don't mind waiting, I can push a tag to do a major version bump and we can fix this as a bug.  If you are in a hurry though, we can leave it as is as a feature with the flag and still default to flattening.

I don't have reliable internet (on my phone right now) so it may take a day or two to get a tag created for the major version bump.

passle-chris commented:
No real hurry. A few days is fine and I think you are right that we should fix as a bug and let those who like the existing behaviour just stick with their older version. I'll edit my repo to remove the use of the setting so it will just automatically keep the build structure.

I've already fixed up the style changes (mostly they were not conscious things, just me letting it do what it wanted). I'll hopefully get it all up before I leave work. :)

And thanks again for your help and feedback.

passle-chris committed:
Removed call to launch debugger that should 100% not have been there.

Removed use of Property to determine the destination directory - it was determined that maintaining directory structure should be the standard.

Changed some explicit declarations to implicit to match style guidelines.

Reduced excessive indent of chained method calls.

Added in some "NotNull" method calls and some contracts to try to stop the build complaining. Not overly familiar with their use so not 100% sure this will satisfy all the contracts but I think it will?

passle-chris committed:
GitHub is complaining - this should stop allow it to realise the call to CreateDirectory will work because what is being passed to it is not null or empty.

passle-chris committed:
When building on github it considers this ensure to be unproven. I'm not sure why but hopefully the slightly weaker contract will satisfy it.

passle-chris committed:
So rather than working on the output of the method I'm just telling it to assume the method is returning something non empty.

passle-chris commented:
So a bit of trial and error to get the contracts working but I think I've now addressed all the issues. I need to do a proper test of it again after the refactoring to remove the property but its getting late here so I'll do that in the morning.

MicahZoltu commented:
I believe I have the tag setup to correctly bump the version.  Let me know when you get a chance to test this and I'll follow-up with a merge.  The next version should be 2.0, if it isn't it means I didn't do the tagging correctly.  :)

MicahZoltu commented on code/FromReferencesTask.cs:
```diff
@@ -87,6 +101,19 @@ public override Boolean Execute()
+			{
+				rootRelativePath = rootRelativePath.Substring(1);
+			}
+			return Path.Combine(LibraryDirectoryFullPath, rootRelativePath);
```
If you change this to:

``` cs
var combinedPath = Path.Combine(...);
Contract.Assume(combinedPath != null);
return combinePath;
```

then you will be able to add a `Contract.Ensures(Contract.Result<String>() != null)` to this method.  This will allow you to remove the `Contract.Assume(!String.IsNullOrEmpty(destinationPath));` from the other method above.

The problem is that `Path.Combine` doesn't have the correct contract on it (that it will never return null).

passle-chris replied:
Ah. Thanks for the tip. I'm not that familiar with code contracts (we don't have them in the codebase I work on) but it definitely strikes me as nicer to have the method ensuring rather than assuming after the method has returned. I suspected it was a problem with Path.Combine just not having contracts but hadn't thought to do the assume just before returning like that.

passle-chris committed:
Slightly nicer contracts using Ensures (and an assumption) in our method rather than an assume on the variable we are sticking the return value into.

passle-chris commented:
I can see there is a V2 in releases now and I think I've addressed the code issues (style and otherwise) that we've discussed so I think it is ready for the merge as long as you are happy.

MicahZoltu commented:
Sorry about not merging this the other day.  I thought I already did.  :/
